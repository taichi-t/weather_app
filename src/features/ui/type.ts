export type UiState = {
  ui: {
    units: 'metric' | 'standard';
  };
};

export default UiState;
