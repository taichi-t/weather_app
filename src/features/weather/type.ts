export type WeatherState = {
  weather: {
    condition: string | undefined;
    temperature: number | undefined;
    loading: boolean;
    error: string | undefined;
  };
};
