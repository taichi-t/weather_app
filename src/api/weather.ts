import Axios, { AxiosInstance } from 'axios';

export type WeatherResponse = {
  coord: { lon: number; lat: number };
  weather: [{ id: number; main: string; description: string; icon: string }];
  base: string;
  main: {
    temp: number;
    feels_like: number;
    temp_min: number;
    temp_max: number;
    pressure: number;
    humidity: number;
  };
  visibility: number;
  wind: { speed: number; deg: number; gust: number };
  rain: { '1h': number };
  clouds: { all: number };
  dt: number;
  sys: {
    type: number;
    id: number;
    country: string;
    sunrise: number;
    sunset: number;
  };
  timezone: number;
  id: number;
  name: string;
  cod: number;
};

type GetRequet<T> = (param: T) => Promise<WeatherResponse>;

class WeatherAPI {
  axios: AxiosInstance;

  constructor() {
    this.axios = Axios.create({
      timeout: 10000, // 10 seconds
    });
  }

  fetchWeather: GetRequet<string> = async (url) => {
    try {
      const res = await this.axios.get<WeatherResponse>(url);
      return res.data;
    } catch (err) {
      return Promise.reject(
        new Error(err.response.data.message || err.message)
      );
    }
  };
}

export const weatherApi = new WeatherAPI();
