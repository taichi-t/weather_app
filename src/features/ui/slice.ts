import { createSlice } from '@reduxjs/toolkit';
import { UiState } from './type';

const initialState: UiState = {
  ui: {
    units: 'standard',
    theme: 'theme-light',
  },
};

const uiSlice = createSlice({
  name: 'ui',
  initialState,
  reducers: {
    setUnits: (state) => {
      state.ui.units = state.ui.units === 'standard' ? 'metric' : 'standard';
    },
    setTheme: (state) => {
      state.ui.theme =
        state.ui.theme === 'theme-light' ? 'theme-dark' : 'theme-light';
    },
  },
});

export default uiSlice.reducer;

export const { setUnits, setTheme } = uiSlice.actions;

export const uiSelector = (state: { uiStore: UiState }): UiState => {
  return state.uiStore;
};
