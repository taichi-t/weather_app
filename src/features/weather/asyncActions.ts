import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export type WeatherResponse = {
  coord: { lon: number; lat: number };
  weather: [{ id: number; main: string; description: string; icon: string }];
  base: string;
  main: {
    temp: number;
    feels_like: number;
    temp_min: number;
    temp_max: number;
    pressure: number;
    humidity: number;
  };
  visibility: number;
  wind: { speed: number; deg: number; gust: number };
  rain: { '1h': number };
  clouds: { all: number };
  dt: number;
  sys: {
    type: number;
    id: number;
    country: string;
    sunrise: number;
    sunset: number;
  };
  timezone: number;
  id: number;
  name: string;
  cod: number;
};

export const fetchWeather = createAsyncThunk<WeatherResponse>(
  'weather/fetchWeather',
  async (_, thunkApi) => {
    const APIKEY = process.env.WEATHER_API_KEY;
    const city = 'vancouver';
    const url = `http://api.openweathermap.org/data/2.5/weather?q=${city}&appid=${APIKEY}`;
    try {
      const res = await axios.get<WeatherResponse>(url);
      return res.data;
    } catch (err) {
      return thunkApi.rejectWithValue(err);
    }
  }
);

export default fetchWeather;
