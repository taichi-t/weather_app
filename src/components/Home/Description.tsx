import * as React from 'react';
import { weatherSelector } from '@/features/weather/slice';
import { useSelector } from 'react-redux';
import { uiSelector } from '@/features/ui/slice';
import returnTempSymbol from '@/util/returnTempSymbol';
import { List } from 'react-content-loader';

const Description: React.FC = () => {
  const { weather } = useSelector(weatherSelector);

  const { ui } = useSelector(uiSelector);
  const renderDescription = () => {
    if (weather.loading) {
      return (
        <List
          backgroundColor="var(--color-paper)"
          foregroundColor="var(--color-skeleton)"
        />
      );
    }
    if (weather.error) {
      return false;
    }
    return (
      <>
        <h2 className="text-secondaryText font-bold">
          feel like {Math.round(weather.data?.main.feels_like as number)}
          {returnTempSymbol(ui.units)}, {weather.data?.weather[0].description}.
        </h2>
        <ul className=" border-l-2 border-key px-4 text-secondaryText text-base flex flex-wrap mt-2">
          <li className="mr-3">
            <i
              className={`wi wi-wind towards-${weather.data?.wind.deg}-deg `}
            />{' '}
            {weather.data?.wind.speed} m/s
          </li>
          <li className="mr-3">
            <i className="wi wi-barometer" /> {weather.data?.main.pressure} hPa
          </li>
          <li className="mr-3">
            <i className="wi wi-humidity" /> {weather.data?.main.humidity} %
          </li>
          <li className="mr-3">Visivility: {weather.data?.visibility} </li>
          <li className="mr-3">
            <i className="wi wi-umbrella" /> {weather.data?.rain?.['1h']} mm/h
          </li>
          <li className="mr-3">
            <i className="wi wi-thermometer" /> {weather.data?.main.temp_max}{' '}
            {returnTempSymbol(ui.units)}
          </li>
          <li>
            <i className="wi wi-thermometer-exterior" />{' '}
            {weather.data?.main.temp_min} {returnTempSymbol(ui.units)}
          </li>
        </ul>
      </>
    );
  };
  return <div className="inner-wrap mt-12">{renderDescription()}</div>;
};

export default Description;
