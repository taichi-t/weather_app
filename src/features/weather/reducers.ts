import { ActionReducerMapBuilder } from '@reduxjs/toolkit';
import { fetchWeather } from './asyncActions';
import { WeatherState } from './type';

export const weatherExtraReducers = (
  builder: ActionReducerMapBuilder<WeatherState>
): void => {
  builder.addCase(fetchWeather.fulfilled, (state, action) => {
    state.weather = {
      data: action.payload,
      loading: false,
      error: undefined,
    };
  });
  builder.addCase(fetchWeather.pending, (state) => {
    state.weather = {
      data: undefined,
      loading: true,
      error: undefined,
    };
  });
  builder.addCase(fetchWeather.rejected, (state, action) => {
    state.weather = {
      data: undefined,
      error: action.error.message,
      loading: false,
    };
  });
};

export default weatherExtraReducers;
