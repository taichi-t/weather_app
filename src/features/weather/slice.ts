import { createSlice } from '@reduxjs/toolkit';
import { weatherExtraReducers } from './reducers';
import { WeatherState } from './type';

const initialState: WeatherState = {
  weather: {
    condition: undefined,
    temperature: undefined,
    loading: true,
    error: undefined,
  },
};

const weatherSlice = createSlice({
  name: 'weather',
  initialState,
  reducers: {},
  extraReducers: weatherExtraReducers,
});

export default weatherSlice.reducer;

export const weatherSelector = (state: {
  weatherStore: WeatherState;
}): WeatherState => {
  return state.weatherStore;
};
