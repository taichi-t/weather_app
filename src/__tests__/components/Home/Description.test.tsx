import * as React from 'react';
import Description from '@/components/Home/Description';
import { render } from '@testing-library/react';
import { mocked } from 'ts-jest/utils';
import { uiSelector } from '@/features/ui/slice';
import { weatherSelector } from '@/features/weather/slice';
import { WeatherResponse } from '@/features/weather/type';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
}));
jest.mock('@/features/ui/slice');
jest.mock('@/features/weather/slice');
jest.spyOn(global.Math, 'random').mockImplementation(() => 0);

mocked(uiSelector).mockReturnValue({
  ui: {
    units: 'standard',
    theme: 'theme-light',
  },
});

describe('<Description>', () => {
  beforeEach(() => {
    mocked(uiSelector).mockClear();
  });

  it('renders when loading', () => {
    mocked(weatherSelector).mockReturnValue({
      weather: {
        data: undefined,
        loading: true,
        error: undefined,
      },
    });
    const tree = render(<Description />);
    expect(tree).toMatchSnapshot();
  });

  it('renders when data is loaded', () => {
    const weatherState: WeatherResponse = {
      coord: { lon: 1, lat: 1 },
      weather: [
        { id: 1, main: 'string', description: 'string', icon: 'string' },
      ],
      base: 'string',
      main: {
        temp: 1,
        feels_like: 1,
        temp_min: 1,
        temp_max: 1,
        pressure: 1,
        humidity: 1,
      },
      visibility: 1,
      wind: { speed: 1, deg: 1, gust: 1 },
      rain: { '1h': 1 },
      clouds: { all: 1 },
      dt: 1,
      sys: {
        type: 1,
        id: 1,
        country: 'string',
        sunrise: 1,
        sunset: 1,
      },
      timezone: 1,
      id: 1,
      name: 'string',
      cod: 1,
    };
    mocked(weatherSelector).mockReturnValue({
      weather: {
        data: weatherState,
        loading: false,
        error: undefined,
      },
    });
    const tree = render(<Description />);
    expect(tree).toMatchSnapshot();
  });

  it('renders when getting error', () => {
    mocked(weatherSelector).mockReturnValue({
      weather: {
        data: undefined,
        loading: false,
        error: 'error',
      },
    });
    const tree = render(<Description />);
    expect(tree).toMatchSnapshot();
  });
});
