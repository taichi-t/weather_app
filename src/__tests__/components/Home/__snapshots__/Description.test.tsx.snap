// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Description> renders when data is loaded 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inner-wrap mt-12"
      >
        <h2
          class="text-secondaryText font-bold"
        >
          feel like 
          1
          ℉
          , 
          string
          .
        </h2>
        <ul
          class=" border-l-2 border-key px-4 text-secondaryText text-base flex flex-wrap mt-2"
        >
          <li
            class="mr-3"
          >
            <i
              class="wi wi-wind towards-1-deg "
            />
             
            1
             m/s
          </li>
          <li
            class="mr-3"
          >
            <i
              class="wi wi-barometer"
            />
             
            1
             hPa
          </li>
          <li
            class="mr-3"
          >
            <i
              class="wi wi-humidity"
            />
             
            1
             %
          </li>
          <li
            class="mr-3"
          >
            Visivility: 
            1
             
          </li>
          <li
            class="mr-3"
          >
            <i
              class="wi wi-umbrella"
            />
             
            1
             mm/h
          </li>
          <li
            class="mr-3"
          >
            <i
              class="wi wi-thermometer"
            />
             
            1
             
            ℉
          </li>
          <li>
            <i
              class="wi wi-thermometer-exterior"
            />
             
            1
             
            ℉
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inner-wrap mt-12"
    >
      <h2
        class="text-secondaryText font-bold"
      >
        feel like 
        1
        ℉
        , 
        string
        .
      </h2>
      <ul
        class=" border-l-2 border-key px-4 text-secondaryText text-base flex flex-wrap mt-2"
      >
        <li
          class="mr-3"
        >
          <i
            class="wi wi-wind towards-1-deg "
          />
           
          1
           m/s
        </li>
        <li
          class="mr-3"
        >
          <i
            class="wi wi-barometer"
          />
           
          1
           hPa
        </li>
        <li
          class="mr-3"
        >
          <i
            class="wi wi-humidity"
          />
           
          1
           %
        </li>
        <li
          class="mr-3"
        >
          Visivility: 
          1
           
        </li>
        <li
          class="mr-3"
        >
          <i
            class="wi wi-umbrella"
          />
           
          1
           mm/h
        </li>
        <li
          class="mr-3"
        >
          <i
            class="wi wi-thermometer"
          />
           
          1
           
          ℉
        </li>
        <li>
          <i
            class="wi wi-thermometer-exterior"
          />
           
          1
           
          ℉
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Description> renders when getting error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inner-wrap mt-12"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="inner-wrap mt-12"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Description> renders when loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inner-wrap mt-12"
      >
        <svg
          aria-labelledby="-aria"
          role="img"
          viewBox="0 0 400 110"
        >
          <title
            id="-aria"
          >
            Loading...
          </title>
          <rect
            clip-path="url(#-diff)"
            height="100%"
            role="presentation"
            style="fill: url(#-animated-diff);"
            width="100%"
            x="0"
            y="0"
          />
          <defs
            role="presentation"
          >
            <clippath
              id="-diff"
            >
              <rect
                height="10"
                rx="3"
                ry="3"
                width="250"
                x="0"
                y="0"
              />
              <rect
                height="10"
                rx="3"
                ry="3"
                width="220"
                x="20"
                y="20"
              />
              <rect
                height="10"
                rx="3"
                ry="3"
                width="170"
                x="20"
                y="40"
              />
              <rect
                height="10"
                rx="3"
                ry="3"
                width="250"
                x="0"
                y="60"
              />
              <rect
                height="10"
                rx="3"
                ry="3"
                width="200"
                x="20"
                y="80"
              />
              <rect
                height="10"
                rx="3"
                ry="3"
                width="80"
                x="20"
                y="100"
              />
            </clippath>
            <lineargradient
              id="-animated-diff"
            >
              <stop
                offset="0%"
                stop-color="var(--color-paper)"
                stop-opacity="1"
              >
                <animate
                  attributeName="offset"
                  dur="1.2s"
                  keyTimes="0; 0.25; 1"
                  repeatCount="indefinite"
                  values="-2; -2; 1"
                />
              </stop>
              <stop
                offset="50%"
                stop-color="var(--color-skeleton)"
                stop-opacity="1"
              >
                <animate
                  attributeName="offset"
                  dur="1.2s"
                  keyTimes="0; 0.25; 1"
                  repeatCount="indefinite"
                  values="-1; -1; 2"
                />
              </stop>
              <stop
                offset="100%"
                stop-color="var(--color-paper)"
                stop-opacity="1"
              >
                <animate
                  attributeName="offset"
                  dur="1.2s"
                  keyTimes="0; 0.25; 1"
                  repeatCount="indefinite"
                  values="0; 0; 3"
                />
              </stop>
            </lineargradient>
          </defs>
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inner-wrap mt-12"
    >
      <svg
        aria-labelledby="-aria"
        role="img"
        viewBox="0 0 400 110"
      >
        <title
          id="-aria"
        >
          Loading...
        </title>
        <rect
          clip-path="url(#-diff)"
          height="100%"
          role="presentation"
          style="fill: url(#-animated-diff);"
          width="100%"
          x="0"
          y="0"
        />
        <defs
          role="presentation"
        >
          <clippath
            id="-diff"
          >
            <rect
              height="10"
              rx="3"
              ry="3"
              width="250"
              x="0"
              y="0"
            />
            <rect
              height="10"
              rx="3"
              ry="3"
              width="220"
              x="20"
              y="20"
            />
            <rect
              height="10"
              rx="3"
              ry="3"
              width="170"
              x="20"
              y="40"
            />
            <rect
              height="10"
              rx="3"
              ry="3"
              width="250"
              x="0"
              y="60"
            />
            <rect
              height="10"
              rx="3"
              ry="3"
              width="200"
              x="20"
              y="80"
            />
            <rect
              height="10"
              rx="3"
              ry="3"
              width="80"
              x="20"
              y="100"
            />
          </clippath>
          <lineargradient
            id="-animated-diff"
          >
            <stop
              offset="0%"
              stop-color="var(--color-paper)"
              stop-opacity="1"
            >
              <animate
                attributeName="offset"
                dur="1.2s"
                keyTimes="0; 0.25; 1"
                repeatCount="indefinite"
                values="-2; -2; 1"
              />
            </stop>
            <stop
              offset="50%"
              stop-color="var(--color-skeleton)"
              stop-opacity="1"
            >
              <animate
                attributeName="offset"
                dur="1.2s"
                keyTimes="0; 0.25; 1"
                repeatCount="indefinite"
                values="-1; -1; 2"
              />
            </stop>
            <stop
              offset="100%"
              stop-color="var(--color-paper)"
              stop-opacity="1"
            >
              <animate
                attributeName="offset"
                dur="1.2s"
                keyTimes="0; 0.25; 1"
                repeatCount="indefinite"
                values="0; 0; 3"
              />
            </stop>
          </lineargradient>
        </defs>
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
